include/pagmo/algorithms/gaco.hpp

```diff
@@ -0,0 +1,988 @@
+        if (m_memory == true) {
+            ++m_counter;
+        }
+
+        // We store some useful variables:
+        const auto &prob = pop.get_problem(); // This is a const reference, so using set_seed for example will not be
+        // allowed
+        auto n_x = prob.get_nx(); // This getter does not return a const reference but a copy of the number of
+        // all the variables
+        auto pop_size = pop.size(); // Population size
+        unsigned count_screen = 1u; // regulates the screen output
+
+        // Note that the number of equality and inequality constraints has to be set up manually in the problem
+        // definition, otherwise PaGMO assumes that there aren't any.
+        auto n_obj = prob.get_nobj();
+        auto n_ec = prob.get_nec();
+        auto n_ic = prob.get_nic();
+        auto n_f = prob.get_nf(); // n_f=prob.get_nobj()+prob.get_nec()+prob.get_nic()
+
+        // Other useful variables are stored:
+        std::vector<vector_double> sol_archive(m_ker, vector_double(1 + n_x + n_f, 1));
+        vector_double omega(m_ker);
+        vector_double prob_cumulative(m_ker);
+        std::uniform_real_distribution<> dist(0, 1);
+        std::normal_distribution<double> gauss{0., 1.};
+
+        // PREAMBLE-------------------------------------------------------------------------------------------------
+        // We start by checking that the problem is suitable for this
+        // particular algorithm.
+
+        if (!pop_size) {
+            pagmo_throw(std::invalid_argument, get_name() + " cannot work on an empty population");
+        }
+        if (prob.is_stochastic()) {
+            pagmo_throw(std::invalid_argument,
+                        "The problem appears to be stochastic " + get_name() + " cannot deal with it");
+        }
+        if (m_gen == 0u) {
+            return pop;
+        }
+        // I verify that the solution archive is smaller or equal than the population size
+        if (m_ker > pop_size) {
+            pagmo_throw(std::invalid_argument,
+                        get_name() + " cannot work with a solution archive bigger than the population size");
+        }
+        if (n_obj != 1u) {
+            pagmo_throw(std::invalid_argument, "Multiple objectives detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+
+        // ---------------------------------------------------------------------------------------------------------
+
+        // No throws, all valid: we clear the logs
+        m_log.clear();
+
+        // Main ACO loop over generations:
+        for (decltype(m_gen) gen = 1u; gen <= m_gen; ++gen) {
+            // At each generation we make a copy of the population into popold
+            population popold(pop);
+
+            auto dvs = pop.get_x(); // note that pop.get_x()[n][k] goes through the different individuals of the
```

Comment:
If I am reading the code correctly, ``dvs`` and ``fit`` are not written to inside the loop. If that's the case, then you can mark them ``const`` and create references to them, so that you avoid copying:

```c++
const auto &dvs = pop.get_x();
const auto &fit = pop.get_f();
```

Reply:
I just read why it is useful to couple `&` with `const` in these cases (here: https://stackoverflow.com/questions/44795308/why-put-const-in-c).. interesting!! thanks